src/extension/deflate.rs

```diff
@@ -218,51 +224,117 @@ impl Extension for Deflate {
         (true, false, false)
     }
 
-    fn decode(&mut self, hdr: &mut Header, data: &mut Option<Data>) -> Result<(), crate::BoxError> {
-        match hdr.opcode() {
-            OpCode::Binary | OpCode::Text if hdr.is_rsv1() && hdr.is_fin() => {}
-            OpCode::Continue if hdr.is_fin() => {}
-            _ => return Ok(())
+    fn decode(&mut self, header: &mut Header, data: &mut BytesMut) -> Result<(), BoxedError> {
+        match header.opcode() {
+            OpCode::Binary | OpCode::Text if header.is_rsv1() => {
+                if !header.is_fin() {
+                    self.await_last_fragment = true;
+                    trace!("deflate: not decoding {}; awaiting last fragment", header);
+                    return Ok(())
+                }
+                trace!("deflate: decoding {}", header)
+            }
+            OpCode::Continue if header.is_fin() && self.await_last_fragment => {
+                self.await_last_fragment = false;
+                trace!("deflate: decoding {}", header)
+            }
+            _ => {
+                trace!("deflate: not decoding {}", header);
+                return Ok(())
+            }
         }
-        if let Some(data) = data {
-            data.bytes_mut().extend_from_slice(&[0, 0, 0xFF, 0xFF]); // cf. RFC 7692, section 7.2.2
-            self.buffer.clear();
-            let mut d = Decompress::new_with_window_bits(false, self.their_max_window_bits);
-            while (d.total_in() as usize) < data.as_ref().len() {
-                let off = d.total_in() as usize;
-                self.buffer.reserve(data.as_ref().len() - off);
-                d.decompress_vec(&data.as_ref()[off ..], &mut self.buffer, FlushDecompress::Sync)?;
+
+        if data.is_empty() {
+            return Ok(())
+        }
+
+        data.extend_from_slice(&[0, 0, 0xFF, 0xFF]); // cf. RFC 7692, section 7.2.2
+
+        self.buffer.clear();
+
+        let mut d = Decompress::new_with_window_bits(false, self.their_max_window_bits);
+        while d.total_in() < as_u64(data.len()) {
+            let off: usize = d.total_in().try_into()?;
+            self.buffer.reserve(data.len() - off);
+            let n = d.total_out();
+            unsafe {
+                // `bytes_mut()` is marked unsafe because it returns a
+                // reference to uninitialised memory. Since we only write
+                // to this memory, usage is safe here.
+                //
+                // `advance_mut()` is marked unsafe because it can not
+                // know if the memory is safe to read. Since we only
+                // advance for as many bytes as we have decompressed,
+                // usage is safe here.
+                d.decompress(&data[off ..], self.buffer.bytes_mut(), FlushDecompress::Sync)?;
```

Comment:
Is a reference to uninitialized memory ever allowed?

Reply:
> Is a reference to uninitialized memory ever allowed?

Yes.

> [..] it may still be worth adjusting the above comment regarding safety since this fact (that the slice is not read from) cannot actually be guaranteed by this method itself?

I have added a dedicated feature `deflate_with_uninitialised_memory` to turn off memory initialisation. Unless it is enabled, memory will be zeroed first (see the newly introduced `initialise` function in this module).

Reply:
A reference to a valid but uninitialized memory location is valid, in the sense that one can create it without adverse effects. However, reading from that memory without prior initialization is UB (even for `u8`).